pr_watch/migrations/0004_auto_20171108_1805.py

```diff
@@ -0,0 +1,38 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.10.7 on 2017-11-08 18:05
+from __future__ import unicode_literals
+
+from django.db import migrations
+
+
+class Migration(migrations.Migration):
+
+    def create_watchedforks(apps, schema_editor):
+        WatchedPullRequest = apps.get_model('pr_watch', 'WatchedPullRequest')
+        WatchedFork = apps.get_model('pr_watch', 'WatchedFork')
+        default_fork = WatchedFork.objects.create(
+            enabled=True,
+            # You will need to replace these values through the shell to make them match your old WATCH_FORK
+            # and WATCH_ORGANIZATION variables (which are not available here because they were removed)
+            fork='edx/edx-platform',
+            organization='edx',
```

Comment:
@pomegranited I agree the data migration is not necessary. But it may be convenient for upgrading our two shared Ocim instances and each developer's instance, to avoid them having to create an entry in the django admin manually. I'm fine either way.

The main things I want to avoid are:
* Avoid having two different places where things are configured (.env and django) - put all Watched Fork settings into the django admin only, and avoid the "create if none exists" approach. Keep behavior simple and predictable.
* Avoid having models that generate new django migrations if some setting gets changed. (Because the setting is used in the model defaults etc.)

> That way by default, Ocim is just an "instance manager", not a "PR sandbox instance manager", until you configure it to watch github.

+1 for that - Ocim should definitely be usable with the whole pr_watch feature disabled.

Reply:
@pomegranited @bradenmacdonald I propose a middle solution to avoid creating WatchedFork, please comment:
- in the migration, create the default WatchedFork object *only when* WatchedPullRequest exist (count()>0). Therefore, by default there'll be 0 WF and 0 WPR, and the feature will be disabled
- keep WPR.watched_fork required, and no default, because from now on it will be required
- do the change in both our 2 Ocim instances manually. The change will involve changing 2 text fields (fork, organization) in the newly-created WatchedFork and clicking Save. The WatchedFork will already be assigned to all WatchedPullRequest. The WPR themselves don't need to be changed
- actually the last step isn't needed. Instead, I suggest that the migration should read `env('WATCH_FORK')` and `env('WATCH_ORGANIZATION')` (note the `env`, not `settings`), so that *if* there's an .env file which defines these values, they will be used in the migration and the manual change won't be needed. The value should be `env('WATCH_FORK', getattr(settings, 'DEFAULT_FORK', 'edx/edx-platform'))`, so that after the developer removes the variables from .env, the migrations still run. No migrations will be created if settings change.

With this solution, there's no manual change to do in our 2 Ocims or developer machines, and we don't need compatibility code to handle nulls+non-nulls.

Reply:
@clemente +1 - that sounds great, and I really like that it allows us to keep WPR.watched_fork required.

Reply:
@clemente Great idea!
However, I'm not sure whether, the existing `WatchedPullRequest` records will automatically get their new `watched_fork` field assigned the new WatchedFork object, unless there's a `default` value provided?  But give it a try and see if it works.

Reply:
Yes they will get `watched_fork` assigned because the migration assigns it to all existing objects.

Reply:
@clemente  But there's a chicken-and-egg problem?
1. `WatchedFork` model gets added
1. Data migration: create initial `WatchedFork` instance if `WatchedPullRequest.objects.count() > 0`
0. `WatchedPullRequest.watched_fork` field gets added, but has no default value, and can't be null.  So it'll error for the existing `WatchedPullRequest` records.

How to assign it a default value that works?

I recommend giving it a try though, I might be wrong about how the defaults get assigned.

Reply:
@pomegranited step 2 not only creates the `WatchedFork`, but also [assigns it](https://github.com/open-craft/opencraft/blob/45548693b93c8999f8b66e8ab3f5588153d4a181/pr_watch/migrations/0004_auto_20171108_1805.py#L34) to all WatchedPullRequest. When creating the next Django asked whether I want a default or whether there was a previous migration to take care of assigning default values (this case).

I implemented the approach I suggested, and it works and creates the WatchedFork with the current values, or a default one if the values are removed from .env.

I must still implement the rest of the changes and have it ready soon.

Reply:
@clemente Excellent, glad your approach worked!